gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/iap/IAPAuthenticationFilter.groovy

```diff
@@ -0,0 +1,204 @@
+
+      // Save the signature to skip validation for subsequent requests with same token.
+      req.getSession().setAttribute(signatureAttribute, jwt.signature)
+
+    } catch (Exception e) {
+      logger.error("Could not verify JWT Token", e)
+      res.sendError(HttpServletResponse.SC_FORBIDDEN, "Could not verify JWT Token")
+      req.getSession(false)?.invalidate()
+      return
+    }
+
+    chain.doFilter(req, res)
+  }
+
+  @Override
+  void init(FilterConfig filterConfig) throws ServletException {}
+
+  @Override
+  void destroy() {}
+
+  boolean isServiceAccount(String email) {
+    if (!email || !permissionService.isEnabled()) {
+      return false
+    }
+    try {
+      def serviceAccounts = front50Service.getServiceAccounts()
+      return serviceAccounts.find { email.equalsIgnoreCase(it.name) }
+    } catch (RetrofitError re) {
+      logger.warn("Could not get list of service accounts.", re)
+    }
+    return false
+  }
+
+  private User verifyJWTAndGetUser(SignedJWT jwt) throws Exception {
+    JWSHeader jwsHeader = jwt.getHeader()
+
+    // Header must have algorithm("alg") and "kid"
+    Preconditions.checkNotNull(jwsHeader.getAlgorithm())
+    Preconditions.checkNotNull(jwsHeader.getKeyID())
+
+    JWTClaimsSet claims = jwt.getJWTClaimsSet()
+
+    // Claims must match audience, issuer
+    Preconditions.checkArgument(claims.getAudience().contains(configProperties.audience))
+    Preconditions.checkArgument(claims.getIssuer().contains(configProperties.issuerId))
+
+    // Claims must have issued at time in the past
+    Date currentTime = new Date()
+    Preconditions.checkArgument(claims.getIssueTime().before(currentTime))
+    // Claims must have expiration time in the future
+    Preconditions.checkArgument(claims.getExpirationTime().after(currentTime))
+
+    // Must have subject, email
+    Preconditions.checkNotNull(claims.getSubject())
+    Preconditions.checkNotNull(claims.getClaim("email"))
+
+    // Verify using public key : lookup with key id, algorithm name provided
+    ECPublicKey publicKey = getKey(jwsHeader.getKeyID(), jwsHeader.getAlgorithm().getName())
+
+    Preconditions.checkNotNull(publicKey)
+    JWSVerifier jwsVerifier = new ECDSAVerifier(publicKey)
+    Preconditions.checkState(jwt.verify(jwsVerifier))
+
+    return new User(email: claims.getClaim("email"))
+  }
+
+  private ECPublicKey getKey(String kid, String alg) throws Exception {
+    JWK jwk = keyCache.get(kid)
+    if (!jwk) {
+      // update cache loading jwk public key data from url
+      JWKSet jwkSet = JWKSet.load(new URL(configProperties.iapVerifyKeyUrl))
+      for (JWK key : jwkSet.getKeys()) {
+        keyCache.put(key.getKeyID(), key)
+      }
+      jwk = keyCache.get(kid)
+    }
+    // confirm that algorithm matches
+    if (jwk != null && jwk.algorithm.name == alg) {
+      return ECKey.parse(jwk.toJSONString()).toECPublicKey()
+    }
+    return null
+  }
+
+  /**
+   * Clears the public key cache every 5 hours.
```

Comment:
This would be a good place for stating _why_ the key is being cleared periodically :-)

Reply:
Done.